stan/math/prim/mat/fun/append_array.hpp

```diff
@@ -1,80 +1,141 @@
-#ifndef STAN_MATH_PRIM_ARR_FUN_APPEND_ARRAY_HPP
-#define STAN_MATH_PRIM_ARR_FUN_APPEND_ARRAY_HPP
+#ifndef STAN_MATH_PRIM_MAT_FUN_APPEND_ARRAY_HPP
+#define STAN_MATH_PRIM_MAT_FUN_APPEND_ARRAY_HPP
 
 #include <stan/math/prim/mat/fun/Eigen.hpp>
-#include <stan/math/prim/scal/meta/return_type.hpp>
+#include <stan/math/prim/mat/meta/append_return_type.hpp>
+#include <stan/math/prim/scal/err/check_consistent_sizes.hpp>
 #include <vector>
 
 namespace stan {
   namespace math {
+    /**
+     * This function copies the value at x to y, casting from the type of T1 to
+     * type T2. This implementation assumes the necessary copy constructors are
+     * available to do the casting from T1 to T2.
+     *
+     * Specializations for std::vector and Eigen::Matrix are provided to copy
+     * and cast the scalar elements.
+     *
+     * @tparam Type of source
+     * @tparam Type of destination
+     * @param x Source variable
+     * @param y Target variable
+     */
+    template<typename T1, typename T2>
+    inline void append_array_copy(const T1& x, T2& y) {
+      y = T2(x);
+    }
 
     /**
-     * Return the concatenation of two specified integer vectors in the
-     *   order of the arguments.
+     * This copies and casts the elements of x to y (by delegating to the Eigen
+     * template cast)
      *
-     * @param x First vector
-     * @param y Second vector
-     * @return A vector of x and y concatenated together (in that order)
+     * @tparam Scalar type of source matrix
+     * @tparam Scalar Type of destination matrix
+     * @tparam R Eigen RowsAtCompileTime of both matrices
+     * @tparam C Eigen ColsAtCompileTime of both matrices
+     * @param x Source matrix
+     * @param y Target matrix
      */
-    inline std::vector<int>
-    append_array(const std::vector<int>& x, const std::vector<int>& y) {
-      std::vector<int> z;
-      z.reserve(x.size() + y.size());
-      z.insert(z.end(), x.begin(), x.end());
-      z.insert(z.end(), y.begin(), y.end());
-      return z;
+    template<typename T1, typename T2, int R, int C>
+    inline void append_array_copy(const Eigen::Matrix<T1, R, C>& x,
+                     Eigen::Matrix<T2, R, C>& y) {
```

Comment:
Second arguments should be lined up under the first unless they won't fit on a line.